### ISSUE TITLE
[BUG] Bettercap

### ISSUE THREAD

**Issue by shashankx86** [aluminum-ice/pwnagotchi#61]
**Describe the bug**
Bettercap not running

Display: false (using webUI)

**Expected behavior**
Run pwnagotchi

**Environment (please complete the following information):**
 - Pwnagotchi: 1.7.9
 - OS version: Linux pwnagotchi 5.10.103-v7l+ #1529 SMP Tue Mar 8 12:24:00 GMT 2022 armv7l GNU/Linux
 - Raspberry pi 4

**Additional context**
Bettercap not running and stuck on "waiting for bettercap API to be available"

```
pi@pwnagotchi:~ $ tail -f /var/log/pwnagotchi.log
[2023-05-03 08:18:14,321] [INFO] waiting for bettercap API to be available ...
[2023-05-03 08:18:15,331] [INFO] waiting for bettercap API to be available ...
[2023-05-03 08:18:16,339] [INFO] waiting for bettercap API to be available ...
[2023-05-03 08:18:17,349] [INFO] waiting for bettercap API to be available ...
[2023-05-03 08:18:18,359] [INFO] waiting for bettercap API to be available ...
```

```
pi@pwnagotchi:~ $ journalctl -fu bettercap
-- Logs begin at Wed 2023-05-03 07:49:02 IST. --
May 03 08:18:47 pwnagotchi systemd[1]: bettercap.service: Scheduled restart job, restart counter is at 56.
May 03 08:18:47 pwnagotchi systemd[1]: Stopped bettercap api.rest service..
May 03 08:18:47 pwnagotchi systemd[1]: Started bettercap api.rest service..
May 03 08:18:47 pwnagotchi bettercap-launcher[3940]: grep: /sys/class/net/mon0/operstate: No such file or directory
May 03 08:18:47 pwnagotchi bettercap-launcher[3940]: command failed: Operation not supported (-95)
May 03 08:18:48 pwnagotchi bettercap-launcher[3940]: WARNING: Terminal colors have been disabled, view will be very limited.
May 03 08:18:48 pwnagotchi bettercap-launcher[3940]: bettercap v2.32.0 (built for linux arm with go1.20.6) [type 'help' for a list of commands]
May 03 08:18:48 pwnagotchi bettercap-launcher[3940]: no interface matching 'mon0' found.
May 03 08:18:48 pwnagotchi systemd[1]: bettercap.service: Main process exited, code=exited, status=1/FAILURE
May 03 08:18:48 pwnagotchi systemd[1]: bettercap.service: Failed with result 'exit-code'.
```

```
pi@pwnagotchi:~ $ sudo ifconfig -a
eth0: flags=4099<UP,BROADCAST,MULTICAST>  mtu 1500
        ether e4:5f:01:0e:b7:79  txqueuelen 1000  (Ethernet)
        RX packets 0  bytes 0 (0.0 B)
        RX errors 0  dropped 0  overruns 0  frame 0
        TX packets 0  bytes 0 (0.0 B)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0

lo: flags=73<UP,LOOPBACK,RUNNING>  mtu 65536
        inet 127.0.0.1  netmask 255.0.0.0
        inet6 ::1  prefixlen 128  scopeid 0x10<host>
        loop  txqueuelen 1000  (Local Loopback)
        RX packets 4093  bytes 261299 (255.1 KiB)
        RX errors 0  dropped 0  overruns 0  frame 0
        TX packets 4093  bytes 261299 (255.1 KiB)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0

usb0: flags=4163<UP,BROADCAST,RUNNING,MULTICAST>  mtu 1500
        inet 10.0.0.2  netmask 255.255.255.0  broadcast 10.0.0.255
        inet6 fe80::6ce0:feff:fe54:8875  prefixlen 64  scopeid 0x20<link>
        ether 6e:e0:fe:54:88:75  txqueuelen 1000  (Ethernet)
        RX packets 6683  bytes 641701 (626.6 KiB)
        RX errors 0  dropped 10  overruns 0  frame 0
        TX packets 5793  bytes 1834576 (1.7 MiB)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0

wlan0: flags=4098<BROADCAST,MULTICAST>  mtu 1500
        ether e4:5f:01:0e:b7:7a  txqueuelen 1000  (Ethernet)
        RX packets 0  bytes 0 (0.0 B)
        RX errors 0  dropped 0  overruns 0  frame 0
        TX packets 0  bytes 0 (0.0 B)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0
```

**Comment by shashankx86** [comment]
tried change user and pass for bettercap and reboot
still same

**Comment by shashankx86** [comment]
tried changing main.iface "mon0" to "wlan0"
restart, reboot 
still getting error with mon0

**Comment by itsNeutron** [comment]
Same here on same hardware.

**Comment by shashankx86** [comment]
> Same here on same hardware.

any be rpi 4 specific issue

**Comment by shashankx86** [comment]
@aluminum-ice help!!

**Comment by shashankx86** [comment]
REMOVED

@aluminum-ice ???

**Comment by aluminum-ice** [comment]
I don't actively support the RPi4. It's pretty clear from the README. If you keep spamming me on this topic I will remove your access. 

Try to fix it and then submit a PR with the fix. I'm not actively spending time on pwnagotchi because my RPiZ2W works the way I want.

**Comment by shashankx86** [comment]
sorry if it caused any inconvenience
can just tell me what "could" be the reason's 
i am completely clueless what to do and whats causing this error

and
Thank you keeping pwnagotchi updated @aluminum-ice

**Comment by shashankx86** [comment]
and that was time i tagged you, will not form now on

**Comment by aluminum-ice** [comment]
It's likely that nexmon isn't working correctly. It's tricky to compile as part of the image build process and I only really spent time getting it right for the 2W. 

Try cloning the nexmon repo directly onto your pwnagotchi and then build nexmon from scratch. I bet you it will start working.

**Comment by sizzlik** [comment]
the settings from the config.toml dont get set correctly. Check the user/pass in /etc/pwnagotchi/config.toml and make sure it is the same in /usr/local/share/bettercap/caplets/pwnagotchi-auto.cap and /usr/local/share/bettercap/caplets/pwnagotchi-manual.cap

For some reason setting you make in the config.toml (no matter if direct sd edit, ssh edit or webcfg) are not setup in other files. Took me quete a while to change my pwnagotchi name. It changed on my ink display, but to connect i still needed pwnagotchi.local and not newname.local..but bettercap needed to connect to newname.api...no clue whats causing it..

**Comment by itsNeutron** [comment]
If the author lets me I’ve found a repo about a pwnagotchi made for RPi4-3 that works out of the box
Inviato da iPhone

Il giorno 8 nov 2023, alle ore 08:58, Moe Sizzlik ***@***.***> ha scritto:

﻿

the settings from the config.toml dont get set correctly. Check the user/pass in /etc/pwnagotchi/config.toml and make sure it is the same in /usr/local/share/bettercap/caplets/pwnagotchi-auto.cap and /usr/local/share/bettercap/caplets/pwnagotchi-manual.cap

For some reason setting you make in the config.toml (no matter if direct sd edit, ssh edit or webcfg) are not setup in other files. Took me quete a while to change my pwnagotchi name. It changed on my ink display, but to connect i still needed pwnagotchi.local and not newname.local..but bettercap needed to connect to newname.api...no clue whats causing it..

**Comment by shashankx86** [comment]
> If the author lets me I’ve found a repo about a pwnagotchi made for RPi4-3 that works out of the box

tell me

**Comment by tubalainen** [comment]
-Retracted-
Based on old RPI Zero hardware that is not compatible with this fork (kernel drivers).

**Comment by sizzlik** [comment]
> Nov 10 10:03:53 pwnagotchi bettercap-launcher[3052]: no interface matching 'mon0' found.

this could be your problem..mon0 isnt your wifi adapter. Check "ip a" or the depricated ifconfig to check the name of you wifi/internet bound adapter. maybe you see something along the line of "wlpxsx" where x is some number

**Comment by tubalainen** [comment]
> > Nov 10 10:03:53 pwnagotchi bettercap-launcher[3052]: no interface matching 'mon0' found.
> 
> this could be your problem..mon0 isnt your wifi adapter. Check "ip a" or the depricated ifconfig to check the name of you wifi/internet bound adapter. maybe you see something along the line of "wlpxsx" where x is some number

100% agreed. Its an older version of Pi Zero. This for does not have support for that chipset (as of now). No wifi interface is detected.

**Comment by shashankx86** [comment]
update:

i was able to get it working with my usb wifi adapter 
(disabled inbuilt wifi)

if anyone knows a way to get wifi (inbuilt) let me know

**Comment by noncenz** [comment]
I'm seeing this issue on a Zero 2W. Bettercap comes up fine if I have the USB interface plugged in but not when I'm tethered over bluetooth. 

Further investigation shows Bettercap-launcher throwing the error `Grep /sys/class/net/eth0/operstate No such file or directory` when switching into manual mode. Due to that error, the launcher script mistakenly runs Bettercap with the _auto_ caplet instead of the _manual_ version. Looks like the bug is arising from[ this changeset](https://github.com/aluminum-ice/pwnagotchi/commit/b9c7cb34ee4778d5df84dcba72dbccd20864a092), which was fixing a previous bug. 

I'm just getting my feet wet in this codebase, hope this helps.

**Comment by noncenz** [comment]
I played around with this a little more and got closer to root cause:

It looks like there is a race condition in __init__.py `restart(mode)` between writing, reading and deleting the auto/manual override flag files. It causes bettercap-launcher to sometimes launch with the auto caplet when it should be the manual one. You can detect the issue with `tail -n 50 /var/log/daemon.log` and observing which interface the Bettercap api comes up on.

This explains why the bug doesn't surface when connected via USB, as presence of the USB interface being up satisfies a different branch in the `is_auto_mode` logic. 

As a hack, in __init__.py function `restart(mode)` you can add `time.sleep(4)` between the two os.system calls that restart the Bettercap and Pwnagotchi services.

My guess here is that the Pwnagotchi service is deleting the flag file before the Bettercap service has a chance to test for it. The os.system calls used here seem to be blocking on this OS but perhaps the error we get earlier in the Bettercap-launcher script is returning control back to python before the end of the script.

**Comment by shashankx86** [comment]
@noncenz can you do PR with this fix

**Comment by noncenz** [comment]
@shashankx86 I'll PR the change today as a temporary fix but really want to get at root cause. You should know that the performance of the Bettercap ui over bluetooth tethering isn't great; separate issue that seems to exist all the way back to the original release.

**Comment by aluminum-ice** [comment]
This issue was temporary patched with PR #76
